### ISSUE TITLE
consistent-hash: support table-granularity vnode number

### ISSUE THREAD

**Issue by xx01cyx** [singularity-data/risingwave#3396]
Currently, we only have one global vnode number, `VIRTUAL_NODE_COUNT`. All vode mappings share the same vnode number. Namely, all vnode mappings in `hash_mapping.rs` have the same length of `VIRTUAL_NODE_COUNT`.

In our newly proposed design, we are to encode vnode into storage key. To resolve the drawbacks brought about by the new design (mainly in parallel range scan of streaming executors), we figure out a workaround -- limiting the number of vnodes to a relatively small amount, say, 256. And vnode number for a specific table could be increased on need (e.g. scale out). 

To enable this, we should support table-granularity vnode number. That is, **each table will have its own total vnode number (i.e. lengths of vnode mappings differ)**.

**Comment by BowenXiao1999** [comment]
Should it be part of catalog or determined dynamiclly be scheduler?

**Comment by xx01cyx** [comment]
> Should it be part of catalog or determined dynamiclly be scheduler?

IMO its's determined by scheduler. Catalog does not need the info, so I think this could be stored just in meta.

**Comment by fuyufjh** [comment]
Catalog needs this info if someone need to scan/lookup that state. I think it's better to also put it in catalog.

**Comment by xx01cyx** [comment]
> Catalog needs this info if someone need to scan/lookup that state (table). I think it's better to also put it in catalog.

If you mean the partition of the table, the info is stored as vnode mapping in `TableCatalog` now. Vnode number could be inferred from vnode mapping.

BTW, @st1page thinks that vnode mapping should be put in `WorkerNodeManager` in frontend instead of `TableCatalog`. We might create another issue for this.

**Comment by st1page** [comment]
> each table will have its own total vnode number

there is a restriction that each table in the same fragment should have the same vnode number. And to be simpler, I propose to let all state tables and operators in the same "create mv" have the same vnode number.

**Comment by st1page** [comment]
> > Catalog needs this info if someone need to scan/lookup that state (table). I think it's better to also put it in catalog.
> 
> If you mean the partition of the table, the info is stored as vnode mapping in `TableCatalog` now. Vnode number could be inferred from vnode mapping.
> 
> BTW, @st1page thinks that vnode mapping should be put in `WorkerNodeManager` in frontend instead of `TableCatalog`. We might create another issue for this.

the mapping is a dynamic parameter in runtime. It will be generated when the cluster starts or restarts and changes when scale-out/scale-in. I don't think we should persist it in catalog or metastore.
c.c. @fuyufjh

**Comment by BugenZhao** [comment]
> If the mv is scaled to the upper bound (which is determined by its vnode number), we have to drop the mv and recreate it to support a larger vnode number.

Will this be too costly, or does it happens very rarely?

**Comment by xx01cyx** [comment]
> Will this be too costly, or does it happens very rarely?

Not frequently, maybe. I'm actually not very sure on this.

**Comment by fuyufjh** [comment]
Just had a discussion with @st1page, theoretically speaking, we agree that:

- the mapping from vnode to CN should be a kind of **transient** (aka. in-memory) info and should not be stored in catalog. 
- Meanwhile, the number of vnodes in a table should be part of catalog, because it is persisted in data (like number of partitions, under our new storage design)

So, for the frontend nodes, how to store the mapping from vnode to CN? Ideally, it would be an in-memory data structure inside of the batch scheduler and obtained from Meta somehow. But this is just the "ideal" way - it looks not very urgent to do the refactoring right now.
